File: backend/utils/units.py
"""Unit conversion helpers for persisted user preferences."""

KG_PER_LB = 0.45359237
CM_PER_IN = 2.54
ML_PER_FL_OZ = 29.5735295625


def kg_to_lb(kg: float) -> float:
    return kg / KG_PER_LB


def lb_to_kg(lb: float) -> float:
    return lb * KG_PER_LB


def cm_to_ft_in(cm: float) -> tuple[int, int]:
    total_inches = cm / CM_PER_IN
    feet = int(total_inches // 12)
    inches = int(round(total_inches - (feet * 12)))
    if inches == 12:
        feet += 1
        inches = 0
    return feet, inches


def ml_to_oz(ml: float) -> float:
    return ml / ML_PER_FL_OZ


def oz_to_ml(oz: float) -> float:
    return oz * ML_PER_FL_OZ

File: frontend/src/utils/units.ts
export type WeightUnit = 'kg' | 'lb';
export type HeightUnit = 'cm' | 'ft';
export type HydrationUnit = 'ml' | 'oz';

const KG_PER_LB = 0.45359237;
const ML_PER_FL_OZ = 29.5735295625;

export function lbToKg(lb: number): number {
  return lb * KG_PER_LB;
}

export function kgToLb(kg: number): number {
  return kg / KG_PER_LB;
}

export function ftInToCm(ft: number, inches: number): number {
  return ft * 30.48 + inches * 2.54;
}

export function cmToFtIn(cm: number): [number, number] {
  const totalInches = cm / 2.54;
  const ft = Math.floor(totalInches / 12);
  let inches = Math.round(totalInches - ft * 12);
  if (inches === 12) {
    inches = 0;
    return [ft + 1, inches];
  }
  return [ft, inches];
}

export function mlToOz(ml: number): number {
  return ml / ML_PER_FL_OZ;
}

export function ozToMl(oz: number): number {
  return oz * ML_PER_FL_OZ;
}

export function round1(value: number): number {
  return Math.round(value * 10) / 10;
}
